### ISSUE TITLE
Question about forecasting ahead to days outside training/validation set

### ISSUE THREAD

**Issue by AlexMRuch** [jdb78/pytorch-forecasting#67]
Hello! First of all, thank you for this wonderful and easy to use library. It's been a joy to work with, and I am really impressed with all of the features you've so kindly added to it for things like easy visualization and evaluation of feature importance!

I am new to deep learning for time series (working on graphs and NLP has been my focus area), and was wondering how I can get the TFT model to generate forecasts ahead to days that are not in the training/validation data? For example, in your example with the stallion data, you produce graphs showing predicted vs observed on the validation set, but how can I get the model to show me predicted trends for data beyond this (e.g., if one wanted to use the model to predict what volume would be in the future)?

Thanks so much for your insight and advice! Really appreciate it!

**Comment by jdb78** [comment]
Thanks for this excellent question! It is absolutely worth answering in the tutorial. 

Because the TFT requires specifying covariates in the future, you need to construct a dataframe that has a time index for those values.

Taking the example from the demand forecasting tutorial. We assume that the variables known in the future stay the same in the next 6 months (particularly prices):
```
# select last 24 months from data (max_encoder_length is 24)
encoder_data = data[lambda x: x.time_idx > x.time_idx.max() - max_encoder_length]

# select last known data point and create decoder data from it by repeating it and incrementing the month
last_data = data[lambda x: x.time_idx == x.time_idx.max()]
decoder_data = pd.concat(
    [last_data.assign(date=lambda x: x.date + pd.offsets.MonthBegin(i)) for i in range(1, max_prediction_length + 1)],
    ignore_index=True,
)

# add time index consistent with "data"
decoder_data["time_idx"] = decoder_data["date"].dt.year * 12 + decoder_data["date"].dt.month
decoder_data["time_idx"] += encoder_data["time_idx"].max() + 1 - decoder_data["time_idx"].min()

# adjust additional time feature(s)
decoder_data["month"] = decoder_data.date.dt.month.astype(str).astype("category")  # categories have be strings

# combine encoder and decoder data
new_prediction_data = pd.concat([encoder_data, decoder_data], ignore_index=True)
```

And now to predict and visualize some examples, simply run:
```
new_raw_predictions, new_x = best_tft.predict(new_prediction_data, mode="raw",return_x=True)

for idx in range(10):  # plot 10 examples
    best_tft.plot_prediction(new_x, new_raw_predictions, idx=idx, show_future_observed=False);
```

**Comment by AlexMRuch** [comment]
Wow, those updates you just pushed are amazing, @jdb78! Thank you so much! This is incredibly helpful!

I'll try to implement them on my own dataset tomorrow morning. Thanks so much for your quick and helpful reply! I really appreciate it!

**Comment by AlexMRuch** [comment]
Just tested the code you posted and so helpfully pushed in the v0.4.1 release! Worked great! Thanks so much, @jdb78!

Now if only I had more time series data for my TFT to train on! 🤣
